### PR TITLE
Deleting Fields: Show error message on template builder

### PR DESCRIPTION
**Pull request by dfuller114**
# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (voduyemi) [inline src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js:54]: Missing custom label for this string.

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.html
Comment (psadie-sfdo) [inline src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.html:12]: What do you think about using `data-name` instead of adding in `data-id`?

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (psadie-sfdo) [inline src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js:68]: How do we feel about one lining guard clauses like this?

`if (isEmpty(this.fieldMapping)) return;`

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (voduyemi) [inline src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:196]: We seem to be missing the required fields error messages now in the template builder. I think we might have scenarios where we have broken field references and missing required fields, in that case we want to prevent saving a template that does not have those fields (`Account1Imported`, `Account1Name`, `Contact1Imported`, `Contact1LastName`)

**Review by voduyemi — CHANGES_REQUESTED** [review]
Looks good Daniel! Minor comments on custom labels and the missing required field feature.

**Comment by melissabarber** [comment]
@dfuller114 @ptewson-sfdo the custom labels look good for me. I left one comment response for Daniel in the custom labels spreadsheet.

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (psadie-sfdo) [inline src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:149]: Could we rename to `sourceObjectFieldsInfo` or `sourceObjectFieldsDescribe` or `sourceObjectFields`?

File: src/labels/CustomLabels.labels
Comment (ptewson-sfdo) [inline src/labels/CustomLabels.labels:6377]: This may be missing from package.xml!

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (ptewson-sfdo) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:149]: I like `sourceObjectFieldsDescribe` - it's descriptive!

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (ptewson-sfdo) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js:68]: I think we technically want to always drop the curly braces in around conditional clauses according to our style guid, but on a personal level this doesn't bother me one bit.

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js]: this was resolved in a later commit. thanks!

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:193]: @voduyemi @psadie-sfdo added the `errors` property to the page level object. Let me know if this is what you were referring to. https://github.com/SalesforceFoundation/NPSP/pull/5845/commits/acb324dcb4b51eddc4188db2561963cb2c057f70 and https://github.com/SalesforceFoundation/NPSP/pull/5845/commits/425bf8fdea66b50b770ac3ff1045678dd5899440

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.html
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.html]: @psadie-sfdo I used `data-id` here because `data-name` is already used in this input component.

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js:71]: I'd prefer sticking with the style guide if we're okay with that, especially if that's the guide we use whenever we move to having a linting config in the repo.

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js]: done!

File: src/labels/CustomLabels.labels
Comment (dfuller114) [reply src/labels/CustomLabels.labels:6377]: added!

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:193]: @voduyemi i created a brand new scratch org off of master. This scenario where you remove an "always required" field like Account1Imported, that's not being blocked in the master branch currently either. so you might double check there. idk if we regressed on that at some point but in the master branch I was able to remove all the fields you listed above from the template without being blocked.

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (psadie-sfdo) [inline src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:486]: Wondering, do we not need any of the information brought over in the `event` object? We only need to be notified that an event was fired?

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (psadie-sfdo) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:193]: Out of the box, technically the only "required" field in template builder is Donation Donor.

**Review by psadie-sfdo — CHANGES_REQUESTED** [review]
Looking good Daniel! I'll do another run once we have the refactors PR merged into main and in here as well

**Comment by psadie-sfdo** [comment]
The `utilInput` component wasn't really touched in your workstream, but I'm seeing an uncaught in promise error in the `wiredGetObjectInfo` method line 177 whenever we have a broken reference. Seems to be happening because we're able to get the object describe info successfully because the `objectApiName` is valid, but then the field we care about isn't being returned and so throws an error when we try to reference its `dataType` property.

What do you think about updating that and maybe just skipping the type setting whenever the field isn't available to us?

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (voduyemi) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:193]: I took a closer look at this, and I think I was wrong, it's working like it should, if you take all those fields off the template, you should get this error message: 
image

File: src/classes/BGE_BatchGiftEntry_UTIL.cls
Comment (psadie-sfdo) [reply src/classes/BGE_BatchGiftEntry_UTIL.cls:163]: Comment moved to #5860

File: src/classes/BGE_DataImportBatchEntry_CTRL.cls
Comment (psadie-sfdo) [reply src/classes/BGE_DataImportBatchEntry_CTRL.cls:164]: Comment moving to #5860

File: src/classes/BGE_BatchGiftEntry_UTIL.cls
Comment (psadie-sfdo) [reply src/classes/BGE_BatchGiftEntry_UTIL.cls:134]: Comment moving to #5860

File: src/classes/FORM_PermissionValidator.cls
Comment (psadie-sfdo) [reply src/classes/FORM_PermissionValidator.cls:81]: Comment moving to #5860

File: src/classes/FORM_PermissionValidator.cls
Comment (psadie-sfdo) [reply src/classes/FORM_PermissionValidator.cls:90]: Comment moving to #5860

File: src/classes/GE_FormRendererService.cls
Comment (psadie-sfdo) [reply src/classes/GE_FormRendererService.cls:142]: Comment moving to #5860

File: src/classes/FORM_PermissionValidator.cls
Comment (psadie-sfdo) [reply src/classes/FORM_PermissionValidator.cls:84]: Comment moving to #5860

File: src/lwc/geFormSection/geFormSection.html
Comment (psadie-sfdo) [reply src/lwc/geFormSection/geFormSection.html:17]: Comment moving to #5860

File: src/lwc/geFormSection/geFormSection.js
Comment (psadie-sfdo) [reply src/lwc/geFormSection/geFormSection.js:247]: Comment moving to #5860

**Comment by psadie-sfdo** [comment]
One additional, potentially unrelated bug, that Victor and I noticed is that the toast messages whenever there's a blocking error now show the tab id in the message instead of the tab label.

File: src/classes/BDI_FieldMapping.cls
Comment (dfuller114) [reply src/classes/BDI_FieldMapping.cls]: @psadie-sfdo Agreed! Added!

File: src/classes/UTIL_Describe.cls
Comment (dfuller114) [reply src/classes/UTIL_Describe.cls]: updated!

File: src/classes/UTIL_Describe.cls
Comment (dfuller114) [reply src/classes/UTIL_Describe.cls]: updated!

File: src/lwc/geFormSection/geFormSection.js
Comment (dfuller114) [reply src/lwc/geFormSection/geFormSection.js:284]: @psadie-sfdo this one was part of the merge from @bdvorachek's branch. This component wasn't edited as part of this PR. I do agree this could be a good candidate for a separate class though.

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.html
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.html]: went ahead and readjusted the whitespacing of the structure under the `<article>` tag

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js:85]: @psadie-sfdo fixed!

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js:6]: fixed!

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js]: @psadie-sfdo I've updated the custom event by renaming it to make it more generic to validation and have also passed a `showError` property there now to show/hide the metadata warning message. You can see that in commits 7b2337455bc9402966dfae573d297fd8f50b0118, 458061ffb35200af9f0952c4d358f9aee79e64ed, and 998a343f7a8656733b6e46854de4024af5001f66

**Comment by dfuller114** [comment]
> The `utilInput` component wasn't really touched in your workstream, but I'm seeing an uncaught in promise error in the `wiredGetObjectInfo` method line 177 whenever we have a broken reference. Seems to be happening because we're able to get the object describe info successfully because the `objectApiName` is valid, but then the field we care about isn't being returned and so throws an error when we try to reference its `dataType` property.
> 
> What do you think about updating that and maybe just skipping the type setting whenever the field isn't available to us?

Yeah that makes sense. Just committed the change! 6c04f294b98957089e017571e74c83248764f216

**Comment by dfuller114** [comment]
> One additional, potentially unrelated bug, that Victor and I noticed is that the toast messages, whenever there's a blocking error, now display the tab id instead of the tab label. So we see "Please review the following: infoTab" instead of "Please review the following: Template Info"

@psadie-sfdo I think I found where that issue is occurring. We were passing the id in the `tabsWithErrors` collection instead of the label. Must have been changed at some point. Pushed the change here bfccf9770885e96f96e985fc37a345f9ada5eeef.

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (psadie-sfdo) [inline src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:547]: We may need to also call this `validate` in the `handleToggleFieldMapping` method when `fieldMapping.Is_Describable === false` so that the validations are re-triggered when a user removes the broken field mapping through the sidebar.

**Review by psadie-sfdo — CHANGES_REQUESTED** [review]
Just a couple more things! I think we're good to go after this

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (psadie-sfdo) [inline src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:542]: We can remove `fieldMapping` as it's no longer used.

File: src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormField/geTemplateBuilderFormField.js:9]: fixed!

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:549]: fixed!

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (dfuller114) [reply src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js]: fixed!

**Review by psadie-sfdo — APPROVED** [review]
Approved!

**Review by voduyemi — APPROVED** [review]
Looks good to me! @dfuller114. Thank you!

File: src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js
Comment (psadie-sfdo) [inline src/lwc/geTemplateBuilderFormFields/geTemplateBuilderFormFields.js:546]: @ptewson-sfdo - saw your message in slack and looked into it. I think we should wrap this line in a simple `if (element)` so that we don't attempt to uncheck the checkbox for form fields with deleted mappings. They won't be checked onload in any case. Long term I think we should really be disabling these checkboxes in the sidebar for broken field mappings.

**Review by ptewson-sfdo — APPROVED** [review]
Looks good! Thanks for the great effort!

**Pull request merged**
